crates/starfish/core/src/block_manager/block_suspender.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+use std::{
+    collections::{BTreeMap, BTreeSet, btree_map::Entry},
+    sync::Arc,
+    time::Instant,
+};
+
+use itertools::Itertools;
+use starfish_config::AuthorityIndex;
+use tracing::debug;
+
+use crate::{BlockHeaderAPI, BlockRef, VerifiedBlockHeader, context::Context};
+
+struct SuspendedBlockHeader {
+    block_header: VerifiedBlockHeader,
+    unresolved_ancestors: BTreeSet<BlockRef>,
+    timestamp: Instant,
+}
+
+impl SuspendedBlockHeader {
+    fn new(block: VerifiedBlockHeader, missing_ancestors: &BTreeSet<BlockRef>) -> Self {
+        let mut ma = BTreeSet::new();
+        for ancestor in missing_ancestors {
+            ma.insert(*ancestor);
+        }
+        Self {
+            block_header: block,
+            unresolved_ancestors: ma,
+            timestamp: Instant::now(),
+        }
+    }
+}
+
+pub(crate) struct BlockSuspender {
+    context: Arc<Context>,
+    /// Keeps all the suspended block headers. A suspended block header is a
+    /// header that is missing part of its causal history and thus can't be
+    /// immediately processed. A block header will remain in this map until
+    /// all its causal history has been successfully processed.
+    suspended_block_headers: BTreeMap<BlockRef, SuspendedBlockHeader>,
+    /// A map that keeps all the blocks that we are missing (keys) and the
+    /// corresponding blocks that reference the missing blocks as ancestors
+    /// and need them to get unsuspended. It is possible for a missing
+    /// dependency (key) to be a suspended block, so the block has been
+    /// already fetched but itself is still missing some of its ancestors to be
+    /// processed.
+    unresolved_ancestors: BTreeMap<BlockRef, BTreeSet<BlockRef>>,
+    /// A map of currently missing blocks to the set of authorities expected
+    /// to have them available locally. This set is approximated based on the
+    /// block's author and the authors of its direct children.
+    /// A block is considered missing if it appears in `missing_ancestors`
+    /// and has not yet been accepted or fetched. Blocks already stored or
```

Comment:
Can a block be in 'missing_ancestors' and accepted simultaneously? I think not.
```suggestion
    /// A block is considered missing if it appears in `missing_ancestors`
    /// and has not yet been fetched. Blocks already stored or
```

Reply:
resolved in latest push